mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -3138,6 +3138,117 @@ _PUBLIC_ struct emsmdbp_object *emsmdbp_object_ftcontext_init(TALLOC_CTX *mem_ct
+		xml = talloc_asprintf(local_mem_ctx, "TargetRecipients=\"");
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+		emsmdbp_fill_prop_index(&prop_index, msg_data->columns);
+		for (i = 0; i < msg_data->recipients_count; i++) {
+			if (prop_index.email_address != (uint32_t) -1) {
+				xml = talloc_asprintf_append(xml, "%s", (char *)msg_data->recipients[i].data[prop_index.email_address]);
+				OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+				if (i + 1 < msg_data->recipients_count) {
+					/* ; separated */
+					xml = talloc_asprintf_append(xml, ";");
+					OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+				}
+			}
+		}
+		xml = talloc_asprintf_append(xml, "\"");
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	} else {
+		/* This shouldn't happen, but just in case */
+		xml = talloc_asprintf(local_mem_ctx, "");
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	}
+
+	*target_recipients_attr = talloc_strdup(mem_ctx, xml);
+	OPENCHANGE_RETVAL_IF(!*target_recipients_attr, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	talloc_free(local_mem_ctx);
+
+	return MAPI_E_SUCCESS;
+}
+
+/**
+   \details Store the Providers node into providers_node out parameter.
+
+   \param emsmdbp_ctx pointer to the emsmdb provider cotnext
+   \param sharing_object the sharing message object to get the XML data from
+   \param retvals the array of return codes from get_properties
+   \param data_pointers the array of properties values from get_properties of the sharing_object
+   \param include_folder mark if the providers node is for opening a folder
+   \param target_recipients the string where the recipients' email addresses separated by semi colons are stored
+   \param [out] providers_node pointer to string where the Providers XML node is stored
+   \param mem_ctx pointer to memory context where the providers_node is created
+
+   \note On success providers_node string is allocated in mem_ctx. It is up to the developer
+   to free that memory context when it is not needed anymore.
+
+   \return MAPI_E_SUCCESS on success, a mapi error code otherwise
+ */
+static enum MAPISTATUS emsmdbp_object_sharing_metadata_providers_node(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *sharing_object, enum MAPISTATUS *retvals, void **data_pointers, bool include_folder, char *target_recipients, char **providers_node, TALLOC_CTX *mem_ctx)
+{
+	char	       *xml;
+	TALLOC_CTX     *local_mem_ctx;
+
+	local_mem_ctx = talloc_named(NULL, 0, "sharing_metadata_providers_node");
+	OPENCHANGE_RETVAL_IF(!local_mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	xml = talloc_asprintf(local_mem_ctx, "<Providers><Provider Type=\"ms-exchange-internal\" %s >", target_recipients);
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	/* PidLidSharingRemoteUid */
+	if (retvals[7] == MAPI_E_SUCCESS && include_folder) {
```

Comment:
retvals[7] is a prone error condition. I have not checked what triggers this function to be called, but I would feel much safer if you either:
- updated the prototype to only rely on PidLidSharingRemoteUid and PidLigSharingRemoteStoreUid rather than the whole properties and retvals set
- reliably extracted this information at the beginning of the function